sql-plugin/src/main/340/scala/com/nvidia/spark/rapids/SparkShims.scala

```diff
@@ -64,4 +64,11 @@ object SparkShimImpl extends Spark331PlusShims {
   // AnsiCast is removed from Spark3.4.0
   override def ansiCastRule: ExprRule[_ <: Expression] = null
 
+  override def ignoreTimeZone(e: Expression): Expression = e match {
```

Comment:
I think this method will be required for versions >=340 assuming this doesn't change, so it would be better to move to 340+ shim. Also, I am not sure if this needs to be in SparkShims (i,e should this method override from base trait SparkShims).
@tgravescs , @gerashegalov  - Could you please let us know if there are any guidelines on when a method should be declared inside SparkShims(trait SparkShims) vs creating a method inside a different object(in a separate file - ParquetStringPredShims)

Reply:
SparkShims used to be a kitchen sink for all use cases before Spark 3.2. It does not scale physically nor logically. But I am not in favor of doing tiny classes/objects either, especially if we end up following unnecessary pattern of a single class per file.

I think we need larger themes like DataSourceShims, ExpressionShims, need to see what makes sense . Always consider common base traits to reduce code duplication. E.g.  ExpressionShims31X, ExpressionShims32X to be extended by a concrete class/object ExpressionShims321 with a minimum delta.

Reply:
Okay never mind it is the `needsTimeZone` that is the issue. Could we just shim that part?

Reply:
I was trying to retain the functionality while removing the reference to `CastBase` -- if it's ok to capture all `TimeZoneAwareExpression` expressions we don't even need to shim it.

Reply:
Sorry I was not as explicit here. The question is how much code do we want or need to shim and still keep things clean.

What you have right now with shimming `ignoreTimeZone` is fine to me. It is clear and fairly concise.

If we want to not replicate the `GpuCast` rule in each shim then we have a few options. Note that this is a small amount of code and not likely to change so I am fine with the replication.

We could use `unapply` to do a clean version of the code you showed.
https://docs.scala-lang.org/tour/extractor-objects.html

```scala
object CpuCastWithoutTimezone {
  def unapply(exp: Expression): Option[TimeZoneAwareExpression] = exp match {
    case c: CastBase if c.timeZoneId.nonEmpty && !c.needsTimeZone => Some(c)
    case _ => None
  }
}

...

  def ignoreTimeZone(e: Expression): Expression = e match {
    case c: CpuCastWithoutTimezone => c.withTimeZone(null)
    case c: GpuCast if c.timeZoneId.nonEmpty && !c.needsTimeZone =>
      c.withTimeZone(null)
    case _ => e
  }

```

For me either solution is fine. The latter is likely going to be more code for each shim, but a clean separation of CPU and GPU code.

Reply:
Experimented with this approach and I find the original approach much more straightforward (we haven't used the unapply approach in other cases).

Reply:
I think one of @revans2's comments was giving a good direction that we can get away without introducing a shim for this logic.

We already utilize in GpuCast that c.needsTimezone actually just delegates to the object method Cast.needsTimeZone(child.dataType, dataType). Moreover GpuCast also mixes in TimeZoneAwareExpression 

Thus the unhshimmed version could  be

```scala
  private def ignoreTimeZone(e: Expression): Expression = e match {
    case c: TimeZoneAwareExpression
      if c.prettyName.contains("Cast") &&
        c.timeZoneId.nonEmpty &&
        !Cast.needsTimeZone(c.children.head.dataType, c.dataType) =>
      c.withTimeZone(null)
    case _ => e
  }
```

If we want to keep a more type-safe shim version I suggest to stick it into one of the shim objects 
under sql-plugin/src/main/311until340-all/ otherwise we triplicate the same logic currently